src/jit/importer.cpp

```diff
@@ -5953,6 +5953,117 @@ GenTree* Compiler::impImportLdvirtftn(GenTree*                thisPtr,
     return gtNewHelperCallNode(CORINFO_HELP_VIRTUAL_FUNC_PTR, TYP_I_IMPL, helpArgs);
 }
 
+//------------------------------------------------------------------------
+// impBoxPatternMatch: match and import common box idioms
+//
+// Arguments:
+//   pResolvedToken - resolved token from the box operation
+//   codeAddr - position in IL stream adfter to mach
+//   codeEndp - end of IL stream to match
+//
+// Return Value:
+//   Number of IL bytes matched and imported, -1 otherwise
+//
+
+int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken, const BYTE* codeAddr, const BYTE* codeEndp)
+{
+    if (codeAddr >= codeEndp)
+        return -1;
+
+    switch (codeAddr[0])
+    {
+    case CEE_UNBOX_ANY:
+        // box + unbox.any
+        if (codeAddr + 1 + sizeof(mdToken) <= codeEndp)
+        {
+            CORINFO_RESOLVED_TOKEN unboxResolvedToken;
+
+            impResolveToken(codeAddr + 1, &unboxResolvedToken, CORINFO_TOKENKIND_Class);
+
+            // See if the resolved tokens describe types that are equal.
+            const TypeCompareState compare =
+                info.compCompHnd->compareTypesForEquality(unboxResolvedToken.hClass, pResolvedToken->hClass);
+
+            // If so, box/unbox.any is a nop.
+            if (compare == TypeCompareState::Must)
+            {
+                JITDUMP("\n Importing BOX; UNBOX.ANY as NOP\n");
+                // Skip the next unbox.any instruction
+                return 1 + sizeof(mdToken);
+            }
+        }
+        break;
+
+    case CEE_BRTRUE:
+    case CEE_BRTRUE_S:
+    case CEE_BRFALSE:
+    case CEE_BRFALSE_S:
+        // box + br_true/false
+        if ((codeAddr + ((codeAddr[0] >= CEE_BRFALSE) ? 5 : 2)) <= codeEndp)
+        {
+            if (!(impStackTop().val->gtFlags & GTF_SIDE_EFFECT))
```

Comment:
Would it be better to just spill the side-effects here instead?

Reply:
I  think it's cleaner like you have it -- leaves the eval stack unmodified when the patterns don't match.